src/DynamoCore/Extensions/LinterExtensionBase.cs

```diff
@@ -0,0 +1,274 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using Dynamo.Graph.Nodes;
+using Dynamo.Graph.Workspaces;
+using Dynamo.Linting;
+using Dynamo.Linting.Rules;
+
+namespace Dynamo.Extensions
+{
+    /// <summary>
+    /// Base class for all LinterExtensions
+    /// </summary>
+    public abstract class LinterExtensionBase : IExtension
+    {
+        private const string NODE_ADDED_PROPERTY = "NodeAdded";
+        private const string NODE_REMOVED_PROPERTY = "NodeRemoved";
+
+        #region Private/Internal properties
+        private HashSet<LinterRule> linterRules = new HashSet<LinterRule>();
+        private LinterManager linterManager;
+        private WorkspaceModel currentWorkspace;
+
+        internal ReadyParams ReadyParamsRef { get; set; }
+
+        internal bool IsActive => this.linterManager?.IsExtensionActive(UniqueId) ?? false;
+  
+        internal LinterExtensionDescriptor ExtensionDescriptor { get; private set; }
+        #endregion
+
+        #region Public properties
+
+        ///<inheritdoc/>
+        public abstract string UniqueId { get; }
+
+        ///<inheritdoc/>
+        public abstract string Name { get; }
+
+        /// <summary>
+        /// Collection of the rules in this extension
+        /// </summary>
+        public HashSet<LinterRule> LinterRules => linterRules;
+
+        #endregion
+
+        #region Internal methods
+
+        /// <summary>
+        /// Add a LinterRule
+        /// </summary>
+        /// <param name="linterRule"></param>
+        public void AddLinterRule(LinterRule linterRule)
+        {
+            linterRules.Add(linterRule);
+        }
+
+        /// <summary>
+        /// Remove a LinterRule
+        /// </summary>
+        /// <param name="linterRule"></param>
+        public void RemoveLinterRule(LinterRule linterRule)
+        {
+            if (linterRules.Contains(linterRule))
```

Comment:
I think we can call Remove directly, it will return false if it doesn't contain the rule and we avoid one search.